src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/alerts/AlertsReconciliationService.kt

```diff
@@ -0,0 +1,94 @@
+package uk.gov.justice.digital.hmpps.prisonertonomisupdate.alerts
+
+import com.microsoft.applicationinsights.TelemetryClient
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.withContext
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.config.trackEvent
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.ActivePrisonerId
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.NomisApiService
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.asPages
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.doApiCallWithRetries
+
+@Service
+class AlertsReconciliationService(
+  private val telemetryClient: TelemetryClient,
+  private val dpsAlertsApiService: AlertsDpsApiService,
+  private val nomisAlertsApiService: AlertsNomisApiService,
+  private val nomisApiService: NomisApiService,
+  @Value("\${reports.alerts.reconciliation.page-size}")
+  private val pageSize: Long = 20,
+) {
+  private companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  suspend fun generateReconciliationReport(activePrisonersCount: Long): List<MismatchAlerts> {
+    return activePrisonersCount.asPages(pageSize).flatMap { page ->
+      val activePrisoners = getActivePrisonersForPage(page)
+
+      withContext(Dispatchers.Unconfined) {
+        activePrisoners.map { async { checkActiveAlertsMatch(it) } }
```

Comment:
So how many threads will be running at the same time?

Reply:
And does Unconfined mean you don't have structured concurrency so can handle API call failures?

Reply:
Oh ignore the above, I see now it's just the same pattern as works with other domains.

Reply:
Yep - I can't even remember why any of this was chosen